crates/engine/tree/src/tree/precompile_cache.rs

```diff
@@ -0,0 +1,206 @@
+
+use alloy_primitives::Bytes;
+use reth_evm::precompiles::{DynPrecompile, Precompile};
+use revm::precompile::{PrecompileError, PrecompileResult};
+use std::sync::Arc;
+
+/// Cache for precompiles, for each input stores the result.
+#[derive(Debug, Clone)]
+pub struct PrecompileCache(
+    Arc<mini_moka::sync::Cache<CacheKey, CacheEntry, alloy_primitives::map::DefaultHashBuilder>>,
+);
+
+impl Default for PrecompileCache {
+    fn default() -> Self {
+        Self(Arc::new(
+            mini_moka::sync::CacheBuilder::new(100_000)
+                .build_with_hasher(alloy_primitives::map::DefaultHashBuilder::default()),
+        ))
+    }
+}
+
+impl PrecompileCache {
+    fn get(&self, key: &CacheKey) -> Option<CacheEntry> {
+        self.0.get(key)
+    }
+
+    fn insert(&self, key: CacheKey, value: CacheEntry) {
+        self.0.insert(key, value);
+    }
+
+    fn weighted_size(&self) -> u64 {
+        self.0.weighted_size()
+    }
+}
+
+/// Cache key, just input for each precompile.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct CacheKey(alloy_primitives::Bytes);
+
+/// Combined entry containing both the result and gas bounds.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct CacheEntry {
+    /// The actual result of executing the precompile.
+    result: PrecompileResult,
+    /// Observed gas used on successful run
+    gas_used: u64,
+    /// Gas limit observed on call error.
+    err_gas_limit: u64,
+}
+
+/// A cache for precompile inputs / outputs.
+#[derive(Debug)]
+pub(crate) struct CachedPrecompile {
+    /// Cache for precompile results and gas bounds.
+    cache: PrecompileCache,
+    /// The precompile.
+    precompile: DynPrecompile,
+    /// Cache metrics.
+    metrics: CachedPrecompileMetrics,
+}
+
+impl CachedPrecompile {
+    /// `CachedPrecompile` constructor.
+    pub(crate) fn new(precompile: DynPrecompile, cache: PrecompileCache) -> Self {
+        Self { precompile, cache, metrics: Default::default() }
+    }
+
+    pub(crate) fn wrap(precompile: DynPrecompile, cache: PrecompileCache) -> DynPrecompile {
+        let wrapped = Self::new(precompile, cache);
+        move |data: &Bytes, gas_limit: u64| -> PrecompileResult { wrapped.call(data, gas_limit) }
+            .into()
+    }
+
+    fn increment_by_one_precompile_cache_hits(&self) {
+        self.metrics.precompile_cache_hits.increment(1);
+    }
+
+    fn increment_by_one_precompile_cache_misses(&self) {
+        self.metrics.precompile_cache_misses.increment(1);
+    }
+
+    fn increment_by_one_precompile_errors(&self) {
+        self.metrics.precompile_errors.increment(1);
+    }
+
+    fn cache_size(&self) -> u64 {
+        self.cache.weighted_size()
+    }
+
+    fn update_precompile_cache_size(&self, previous_size: u64) {
+        let new_size = self.cache.weighted_size();
+        if new_size > previous_size {
+            self.metrics.precompile_cache_size.increment((new_size - previous_size) as f64);
```

Comment:
should we just do `set` here with `new_size`?

Reply:
yep makes sense given that it is a gauge, done